linera-views/src/backends/lru_caching.rs

```diff
@@ -50,54 +298,265 @@ static NUM_CACHE_SUCCESS: LazyLock<IntCounterVec> = LazyLock::new(|| {
-            max_cache_size,
+            lru_cache_policy,
+            total_size: 0,
         }
     }
 
-    /// Inserts an entry into the cache.
-    pub fn insert(&mut self, key: Vec<u8>, value: Option<Vec<u8>>) {
-        match self.map.entry(key.clone()) {
+    fn get_lower_bound(&self, key: &[u8]) -> Option<(&Vec<u8>, &FindCacheEntry)> {
+        match self.map_find.range(..=key.to_vec()).next_back() {
+            None => None,
+            Some((key_store, value)) => {
+                if key.starts_with(key_store) {
+                    Some((key_store, value))
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    fn get_lower_bound_mut(&mut self, key: &[u8]) -> Option<(&Vec<u8>, &mut FindCacheEntry)> {
+        match self.map_find.range_mut(..=key.to_vec()).next_back() {
+            None => None,
+            Some((key_store, value)) => {
+                if key.starts_with(key_store) {
+                    Some((key_store, value))
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// Clearing entries of the queue if too long
+    fn insert_queue(&mut self, full_key: (Vec<u8>, CacheEntry), cache_size: usize) {
+        self.queue.insert(full_key, cache_size);
+        self.total_size += cache_size;
+        loop {
+            if self.total_size <= self.lru_cache_policy.max_cache_size
+                && self.queue.len() <= self.lru_cache_policy.max_cache_entries
+            {
+                break;
+            }
+            let Some(value) = self.queue.pop_front() else {
+                break;
+            };
+            match value.0 {
+                (v, CacheEntry::Find) => {
+                    self.map_find.remove(&v);
+                }
+                (v, CacheEntry::Value) => {
+                    self.map_value.remove(&v);
+                }
+            }
+            self.total_size -= value.1;
+        }
+    }
+
+    /// Remove an entry from the queue
+    fn remove_from_queue(&mut self, full_key: &(Vec<u8>, CacheEntry)) {
+        let existing_cache_size = self.queue.remove(full_key).unwrap();
+        self.total_size -= existing_cache_size;
+    }
+
+    /// Set a size to zero in the queue
+    fn zero_set_entry_in_queue(&mut self, full_key: &(Vec<u8>, CacheEntry)) {
+        let len = full_key.0.len();
+        let cache_size = self.queue.get_mut(full_key).unwrap();
+        self.total_size -= *cache_size;
+        *cache_size = len;
+        self.total_size += *cache_size;
+    }
+
+    /// Inserts a value entry into the cache.
+    pub fn insert_value(&mut self, key: Vec<u8>, cache_entry: ValueCacheEntry) {
+        let cache_size = cache_entry.size() + key.len();
+        if cache_size > self.lru_cache_policy.max_cache_size {
+            // Inserting that entry would lead to complete clearing of the cache
+            // which is counter productive
+            return;
+        }
```

Comment:
Isn't it common practice to in general not cache objects that are too big? IMHO I think that no cached item should be bigger than probably like 10% (being generous) of the total size of the cache. I don't think only checking for objects that are bigger than the full size of the cache is enough.
If the objects happen to be too big for the size of the cache (even if they're smaller than the max cache size), in the worse case the cache will be useless, as every new insertion will basically evict the entire existing cache.
And do we know how big what we're caching here will generally be? It might be worth having a metric for it, so we can adjust the size of the cache accordingly if needed.

Reply:
I added a control value `max_entry_size` in the code.